fix(updater): fallback if tmp is on different mount point, closes #4500

lucasfernog opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [x] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [ ] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### Other information

lucasfernog committed:
fix(updater): fallback if tmp is on different mount point, closes #4500

Beanow commented on core/tauri/src/updater/core.rs:
```diff
@@ -640,14 +635,28 @@ impl<R: Runtime> Update<R> {
+
+  // fallback to cache_dir or CWD if the tmp directory is not in the same mount point
+  if extract_path_metadata.dev() != tmp_dir_metadata.dev() {
+    let fallback_dir =
+      dirs_next::cache_dir().unwrap_or_else(|| extract_path.parent().unwrap().to_path_buf());
+    tmp_dir = tempfile::Builder::new()
+      .prefix("tauri_current_app")
+      .tempdir_in(fallback_dir)?;
+    tmp_dir_metadata = tmp_dir.path().metadata()?
```
It doesn't look like we're checking the metadata again to be sure those fallbacks are on the same mountpoint?
Especially if `~/.cache` is found and preferred over the current `.appimage` path, this seems quite possible.
I could be running the `appimage` from a network share with `~/.cache` on a local btrfs volume.

lucasfernog replied:
You're right. I'll see how I can change that.

Beanow replied:
A loop of candidates maybe :] And it should solve a second issue, namely right now the CWD won't be considered if `dirs_next::cache_dir()` is `Some`.

lucasfernog replied:
done @Beanow code might look weird with the vector of closures, but i think it's the way to remove duplicated code lol

Beanow replied:
Yeah that looks like a good approach to me :ok_hand:

Beanow commented on core/tauri/src/updater/core.rs:
```diff
@@ -661,21 +670,21 @@ fn copy_files_and_run<R: Read + Seek>(archive_buffer: R, extract_path: &Path) ->
+      let path = entry.path()?;
+      if path.extension() == Some(OsStr::new("AppImage")) {
+        // if something went wrong during the extraction, we should restore previous app
+        if let Err(err) = entry.extract(extract_path) {
+          Move::from_source(tmp_app_image).to_dest(extract_path)?;
+          return Err(crate::api::Error::Extract(err.to_string()));
+        }
```
Since you found a nice option to compare the mountpoints, the ultimate fallback imo should be to avoid move, and copy across the filesystem boundaries instead.

lucasfernog replied:
well I didn't want to change the logic that much specially since it was audited already. I think if we fallback to the current appimage directory we're safe.

Beanow replied:
That is, if none of the 3 dir options appear to be on the same mount point.

Beanow replied:
> well I didn't want to change the logic that much specially since it was audited already. I think if we fallback to the current appimage directory we're safe.

Got it, sure that makes sense. Because I would _expect_ a sibling to the appimage to be the same file system.
Though we should still check that (previous comment).

lucasfernog replied:
Yeah I did that though it sounds unlikely that it would happen.

Beanow replied:
Current solution looks like a fair middle ground, trying three candidates to move or a clear error message :+1:

lucasfernog commented:
I also removed the `#[cfg(feature = "updater")]`in the core.rs file since we do not have an specific feature for updater documentation generation anymore (we previously had one because the updater feature enabled the zip crate, which didn't cross compile for docs.rs - now fixed by removing its default features).

lucasfernog committed:
feat: enhance fallback

Beanow commented:
I'll leave the call up to a maintainer, but my thoughts are addressed :ok_hand:

lucasfernog commented:
Thanks for the review @Beanow
